### PR TITLE
HPCC-16332 Send error message if root authorization fails

### PR DESCRIPTION
**Pull request by wangkx**
If a user logins to ECLWatch with valid credential without SMC
Root access, the existing ECLWatch is redirected to the login
window as if a bad password is entered. In this fix, the
redirect is replaced by sending an error message.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/bindings/http/platform/httpbinding.cpp [hunk 1]
```diff
@@ -603,27 +605,37 @@ bool EspHttpBinding::basicAuth(IEspContext* ctx)
...
         return false;
+    }
 
     bool authenticated = m_secmgr->authorize(*user, rlist, ctx->querySecureContext());
     if(!authenticated)
     {
+        const char *desc = nullptr;
         if (user->getAuthenticateStatus() == AS_PASSWORD_EXPIRED || user->getAuthenticateStatus() == AS_PASSWORD_VALID_BUT_EXPIRED)
-            ctx->AuditMessage(AUDIT_TYPE_ACCESS_FAILURE, "Authentication", "ESP password is expired");
+            desc = "ESP password is expired";
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpbinding.cpp]: Should be able to call ctx->setRespMsg() and no need for "desc" pointer

File: esp/bindings/http/platform/httpservice.cpp [hunk 2]
```diff
@@ -1230,6 +1235,33 @@ EspAuthState CEspHttpServer::authNewSession(EspAuthRequest& authReq, const char*
...
+        if (isEmptyString(errMsg))
+            resp.append("\"Error\": \"Access Denied.\"");
+        else
+            resp.appendf("\"Error\": \"%s\"", errMsg);
+        resp.append(" }");
+        resp.append(" }");
+    }
+    else
+    {
+        resp.setf("<LoginResponse><Error>");
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: Why setf() ?

File: esp/bindings/http/platform/httpservice.cpp [hunk 3]
```diff
@@ -1230,6 +1235,33 @@ EspAuthState CEspHttpServer::authNewSession(EspAuthRequest& authReq, const char*
...
+    if (format == ESPSerializationJSON)
+    {
+        resp.set("{ ");
+        resp.append(" \"LoginResponse\": { ");
+        if (isEmptyString(errMsg))
+            resp.append("\"Error\": \"Access Denied.\"");
+        else
+            resp.appendf("\"Error\": \"%s\"", errMsg);
+        resp.append(" }");
+        resp.append(" }");
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: Would it make sense to combine some of these appends ?

**Review by RussWhitehead — COMMENTED** [review]
A couple very minor comments

File: esp/bindings/http/platform/httpbinding.cpp (hunk 1 above)
Comment (wangkx) [reply esp/bindings/http/platform/httpbinding.cpp]: The "desc" is also used by the AuditMessage().

File: esp/bindings/http/platform/httpservice.cpp (hunk 2 above)
Comment (wangkx) [reply esp/bindings/http/platform/httpservice.cpp]: I should change it to set().

File: esp/bindings/http/platform/httpservice.cpp (hunk 3 above)
Comment (wangkx) [reply esp/bindings/http/platform/httpservice.cpp]: It shows the JSON layers better to keep those appends.

**Comment by wangkx** [comment]
@RussWhitehead please review again.

**Comment by RussWhitehead** [comment]
Looks good

**Pull request merged**